src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DoNotCallEnumerableCastThatWillFailAnalyzer.cs

```diff
@@ -0,0 +1,171 @@
+                                                                             DiagnosticCategory.Reliability,
+                                                                             RuleLevel.IdeSuggestion,
+                                                                             s_localizableCastDescription,
+                                                                             isPortedFxCopRule: false,
+                                                                             isDataflowRule: false);
+
+        internal static DiagnosticDescriptor OfTypeRule = DiagnosticDescriptorHelper.Create(RuleId,
+                                                                             s_localizableOfTypeTitle,
+                                                                             s_localizableOfTypeMessage,
+                                                                             DiagnosticCategory.Reliability,
+                                                                             RuleLevel.IdeSuggestion,
+                                                                             description: null,
+                                                                             isPortedFxCopRule: false,
+                                                                             isDataflowRule: false);
+
+        private static readonly ImmutableDictionary<string, DiagnosticDescriptor> methodMetadataNames = new Dictionary<string, DiagnosticDescriptor>
+        {
+            ["OfType"] = OfTypeRule,
+            ["Cast"] = CastRule,
+        }.ToImmutableDictionary();
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics
+            => methodMetadataNames.Values.ToImmutableArray();
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+
+            context.RegisterCompilationStartAction(compilationStartContext =>
+            {
+                var enumerableType = compilationStartContext.Compilation.GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemLinqEnumerable);
+                var genericIEnumerableType = compilationStartContext.Compilation.GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemCollectionsGenericIEnumerable1);
+
+                if (enumerableType == null || genericIEnumerableType == null)
+                {
+                    return;
+                }
+
+                compilationStartContext.RegisterOperationAction(operationContext =>
+                {
+                    var invocation = (IInvocationOperation)operationContext.Operation;
+
+                    var targetMethod = invocation.TargetMethod.ReducedFrom ?? invocation.TargetMethod;
+                    if (targetMethod == null || !methodMetadataNames.TryGetValue(targetMethod.Name, out var rule))
```

Comment:
TargetMethod cannot be null
```suggestion
                    if (!methodMetadataNames.TryGetValue(targetMethod.Name, out var rule))
```